pkg/storage/utils/decomposedfs/upload/session.go

```diff
@@ -0,0 +1,269 @@
+	"github.com/cs3org/reva/v2/pkg/appctx"
+	ctxpkg "github.com/cs3org/reva/v2/pkg/ctx"
+	"github.com/cs3org/reva/v2/pkg/logger"
+	"github.com/cs3org/reva/v2/pkg/storage"
+	"github.com/cs3org/reva/v2/pkg/storage/utils/decomposedfs/node"
+	"github.com/cs3org/reva/v2/pkg/utils"
+)
+
+// Session is the interface that OcisSession implements. By combining tus.Upload,
+// storage.UploadSession and custom functions we can reuse the same struct throughout
+// the whole upload lifecycle.
+//
+// Some functions that are only used by decomposedfs are not yet part of this interface.
+// They might be added after more refactoring.
+type Session interface {
+	tusd.Upload
+	storage.UploadSession
+	Persist(ctx context.Context) error
+	Node(ctx context.Context) (*node.Node, error)
+	LockID() string
+	Context(ctx context.Context) context.Context
+	Cleanup(cleanNode, cleanBin, cleanInfo bool)
+}
+
+// OcisSession extends tus upload lifecycle with postprocessing steps.
+type OcisSession struct {
+	store OcisStore
+	// for now, we keep the json files in the uploads folder
+	info tusd.FileInfo
+}
+
+func (s *OcisSession) Context(ctx context.Context) context.Context { // restore logger from file info
+	log, _ := logger.FromConfig(&logger.LogConf{
+		Output: "stderr", // TODO use config from decomposedfs
+		Mode:   "json",   // TODO use config from decomposedfs
+		Level:  s.info.Storage["LogLevel"],
+	})
+	sub := log.With().Int("pid", os.Getpid()).Logger()
+	ctx = appctx.WithLogger(ctx, &sub)
+	ctx = ctxpkg.ContextSetLockID(ctx, s.LockID())
+	return ctxpkg.ContextSetUser(ctx, s.ExecutantUser())
+}
+
+func (s *OcisSession) Purge(ctx context.Context) error {
+	if err := os.Remove(sessionPath(s.store.root, s.info.ID)); err != nil {
+		return err
+	}
+	if err := os.Remove(s.binPath()); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (s *OcisSession) TouchBin() error {
+	file, err := os.OpenFile(s.binPath(), os.O_CREATE|os.O_WRONLY, defaultFilePerm)
+	if err != nil {
+		return err
+	}
+	return file.Close()
+}
+
+func (s *OcisSession) Persist(ctx context.Context) error {
+	uploadPath := sessionPath(s.store.root, s.info.ID)
+	// create folder structure (if needed)
+	if err := os.MkdirAll(filepath.Dir(uploadPath), 0700); err != nil {
+		return err
+	}
+
+	var d []byte
+	d, err := json.Marshal(s.info)
+	if err != nil {
+		return err
+	}
+
+	return os.WriteFile(uploadPath, d, 0600)
+}
+
+func (s *OcisSession) ToFileInfo() tusd.FileInfo {
+	return s.info
+}
+func (s *OcisSession) ProviderID() string {
+	return s.info.MetaData["providerID"]
+}
+func (s *OcisSession) SpaceID() string {
+	return s.info.Storage["SpaceRoot"]
+}
+func (s *OcisSession) NodeID() string {
+	return s.info.Storage["NodeId"]
+}
+func (s *OcisSession) NodeParentID() string {
+	return s.info.Storage["NodeParentId"]
+}
+func (s *OcisSession) NodeExists() bool {
+	return s.info.Storage["NodeExists"] == "true"
+}
+func (s *OcisSession) LockID() string {
```

Comment:
Exported func `LockID` should have a comment or be unexported

Reply:
unexposed